From: Robert <heinro@umich.edu>
Date: Fri, 1 Dec 2017 16:41:58 -0500
Subject: Install python module to Debian-correct path

---
 lib/Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/lib/Makefile b/lib/Makefile
index f15bf60..eb65cee 100644
--- a/lib/Makefile
+++ b/lib/Makefile
@@ -13,7 +13,7 @@ install: install-py
 	mkdir -p -- "$(installdir)"
 	$(INSTALL_DATA) -t $(installdir) $(lib_files)
 
-install-py: installdir = $(DESTDIR)$(prefix)/lib/python3.5/site-packages
+install-py: installdir = $(DESTDIR)$(prefix)/lib/python3/dist-packages
 install-py:
 	$(info Installing python packages ...)
 	for i in $(py_files); do \
