## Script Descriptions
* **Ebot.Output.Extract.Gi.Title.Rev3.py**	-	This script takes a NCBI genbank file, extracts the gi#, the study name, the journal it came from and the sequence and builds a tab delimitted output for each genbank entry.
* **extractGenomes.pl**	-	 extract genomes from NCBI databases (nr/nt) using NCBI Taxon ID, curate and concatenate them to form your own customized database.
* **extractGenbankMetadata.pl**	-	Extract Meta data and protein translations from a Genbank format file.
* **gbk2fna.pl**	-	Read Genbank file and convert it to a Nucleotide Fasta file.
* **getFastaFromAccNos.pl**	
* **getGFF.pl**	-	Given a list of contig names extract GFF data.
* **getGIAnnotation.sh**	
* **getGiInfo.pl**	
* **getGI.pl**	
* **getLineage.pl**	
* **getSciNames.pl**	
* **gff2tbl.pl**	-	Read GFF3(tested with JGI's gff), Contigs and Gene_product files and produce a usable output in NCBI's ridiculous '.tbl' format.
* **GI\_info\_XMLParser.pl**	
